production installer: Added server script to signup for push notifications.

alirizwi opened the pull request:
Added a script scripts/setup/register-server that sends the EXTERNAL_HOST, ZULIP_ADMINISTRATOR, zulip_org_id, and zulip_org_key to an endpoint on zulipchat.com
(/remotes/register_server).

Fixes: #7325

alirizwi commented:
@timabbott How to get values of zulip_org_id and zulip_org_key in the register-server script as these variables are created in a python script generate_secrets.py?

rishig commented on scripts/setup/register-server:
```diff
@@ -0,0 +1,63 @@
+read -p 'Hostname (EXTERNAL_HOST): ' EXTERNAL_HOST
+read -p 'Contact email (ZULIP_ADMINISTRATOR): ' ZULIP_ADMINISTRATOR
+read -p 'Is the information above correct? We recommend using an email address 
+like ops@ that will outlast changes in personnel. [Y/n] ' RESPONSE
```
This section should be moved up before the push notification bouncer text.

rishig replied:
ah, I see where you got this text :). Okay, let's instead just move
```
Use of the bouncer is subject to the Zulip Terms of Service at https://zulipchat.com/terms/.
Do you agree to those terms? [Y/n]
```
to right after 
```
This script will register your server with the Zulip push notification bouncer service.
You can read more about the service at 
    http://zulip.readthedocs.io/en/latest/prod-mobile-push-notifications.html. 
```

Thanks!

alirizwi replied:
@rishig Moved the text.

rishig commented on zilencer/views.py:
```diff
@@ -53,6 +53,18 @@ def remote_server_register_push(request, entity, user_id=REQ(),
 
+@has_request_variables
+def remote_server_register_push_notification(request, hostname, contact_email, uuid, api_key):
+
```
don't need this space

alirizwi replied:
@rishig corrected that. Thanks

rishig commented on zilencer/views.py:
```diff
@@ -53,6 +53,18 @@ def remote_server_register_push(request, entity, user_id=REQ(),
+        contact_email=contact_email,
+        last_updated=timezone.now())
+
+    return json_success()
```
should return failure if not created, right?

alirizwi replied:
@rishig Yes the function should return failure if not created. Added that , Thanks

rishig commented:
Thanks @alirizwi, this is looking great! Adding @gnprice as a reviewer.

gnprice commented on zilencer/views.py:
```diff
@@ -53,6 +53,19 @@ def remote_server_register_push(request, entity, user_id=REQ(),
     return json_success()
 
+@has_request_variables
+def remote_server_register_push_notification(request, hostname, contact_email, uuid, api_key):
```
This name is super confusing -- it's hard to tell how it differs from `remote_server_register_push`, and in fact the answer has nothing to do with the difference in the names.

Let's call it `remote_server_register_server`. Also let's put it above `remote_server_register_push` -- that function logically belongs next to its "unregister" friend, and this function is logically first in that it's what you do before you can do any of the others.

gnprice commented on zilencer/views.py:
```diff
@@ -53,6 +53,19 @@ def remote_server_register_push(request, entity, user_id=REQ(),
+        api_key=api_key,
+        hostname=hostname,
+        contact_email=contact_email,
+        last_updated=timezone.now())
```
Hmm -- I think this says that if there's already one with the right uuid, hostname, etc., but it has a different `last_updated` (which it always will), then we should try to create a new one. Is that right? That doesn't seem like what we mean.

alirizwi replied:
@gnprice fixed that

gnprice commented on scripts/setup/register-server:
```diff
@@ -0,0 +1,62 @@
+zulip_org_id=2
+zulip_org_key=2
+
+if [ "$EXTERNAL_HOST" != "zulip.example.com" ]; then
```
This is backward, isn't it? Ditto below.

gnprice commented on scripts/setup/register-server:
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+EXTERNAL_HOST="zulip.example.com"
+ZULIP_ADMINISTRATOR="zulip-admin@example.com"
```
How do these get set to the real values?

Try using `scripts/get-django-setting`.

alirizwi replied:
I used these value as demo just for manual testing. I dont found anything useful in ```scripts/get-django-setting```. These values are assigned in ```scripts/setup/generate_secrets.py```

alirizwi replied:
I used these value as demo just for manual testing. I dont found anything useful in ```scripts/get-django-setting```. These values are assigned in ```scripts/setup/generate_secrets.py```

alirizwi commented:
@rishig @gnprice Actually the variables zulip_org_id and zulip_org_key used in the script are assigned values in another python script generate_secrets.py. So how do we get those values in the bash script?

alirizwi commented:
@rishig @gnprice I have manually tested the script, it seems to be working now. For the curl request authentication, i have used a random user bot ```outgoing-webhook@zulip.com``` with api key ```d0CpmdtGJEh6ZiozlxJkuivUIsq2NGHy```. Please review it and let me know if any further change is required. Thanks

gnprice commented:
@alirizwi We still need to use real values for EXTERNAL_HOST and ZULIP_ADMINISTRATOR.

The logic inside `remote_server_register_server` is still unnecessarily complicated. For an example of how the logic can go, look at `zilencer/management/commands/add_remote_server.py` (which I added earlier this week.)

Also, please squash your changes into one coherent commit, rather than putting bugfixes on top. Take a look through https://zulip.readthedocs.io/en/latest/contributing/version-control.html .

timabbott commented:
@gnprice it might make sense for you to take over finishing this project, since it seems like an important part of our over installer simplification effort.

alirizwi commented:
@gnprice Can you tell what all is left to be done as its been a long time since this PR is opened. I have added real values for EXTERNAL_HOST and ZULIP_ADMINISTRATOR and squashed the commits. Thanks!

gnprice commented:
@timabbott Yeah, agreed. I'll put the issue (#7325) on my list of things to do in simplifying the installer.

@alirizwi Thanks, your latest version (from yesterday) is an improvement!

It still looks like you have not tested the logic, because it will not work at all. Much like I said in my original review in December, for example, this condition:
```
if [ "$EXTERNAL_HOST" != "zulipdev.com:9991" ]; then
    echo "Please set EXTERNAL_HOST in /etc/zulip/settings.py."
    exit
fi
```
is backward -- it will error out if the admin *has* set an EXTERNAL_HOST.

Closing this PR, because I don't think further rounds of review will be productive.
